Add command line option to enable/disable ResourceLoadStatistics when launching MiniBrowser

mokjasmine opened the pull request:
#### 404718d45a0c05091d7700a11c7368e206eee8c2
<pre>
Add command line option to enable/disable ResourceLoadStatistics when launching MiniBrowser
<a href="https://bugs.webkit.org/show_bug.cgi?id=246233">https://bugs.webkit.org/show_bug.cgi?id=246233</a>

Reviewed by Fujii Hironori.

Use the options --resource-load-statistics-enabled and --resource-load-statistics-disabled to enable and disable ResourceLoadStatistics respectively. It is disabled by default.

* Tools/MiniBrowser/playstation/main.cpp:
(main):
* Tools/MiniBrowser/playstation/MainWindow:
(MainWindow::MainWindow):
(MainWindow::parseOptions):
* Tools/MiniBrowser/playstation/MainWindow.h:

Canonical link: <a href="https://commits.webkit.org/255703@main">https://commits.webkit.org/255703@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87e7fe5c7adf01632ba9bedfd090fa5353a2bc41

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23998 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/103031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2559 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30857 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85724 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99018 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79808 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83702 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37240 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35066 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3941 "Build is in progress. Recent messages:Cleaned up git repository; Checked out pull request; Verified commit is squashed; Reviewed by Fujii Hironori; Validated commit message; Compiled WebKit (warnings); Running layout-tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38940 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1847 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37798 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +115,45 @@ MainWindow::MainWindow(const char* requestedURL)
+
+    for (auto option : options) {
+        bool foundOption = false;
+        __debugbreak();
```
Why do you want to __debugbreak here?. is this intentional?

mokjasmine replied:
resolved in https://github.com/WebKit/WebKit/pull/5154/commits/ec72e16864b4383d8497aac77fa097414823a309

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +115,45 @@ MainWindow::MainWindow(const char* requestedURL)
+
+String MainWindow::executeOption(const char* optionName, const string& option)
+{
+    __debugbreak();
```
Why do you want to __debugbreak here?. is this intentional?

mokjasmine replied:
resolved in https://github.com/WebKit/WebKit/pull/5154/commits/ec72e16864b4383d8497aac77fa097414823a309

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +115,45 @@ MainWindow::MainWindow(const char* requestedURL)
+        __debugbreak();
+        for (auto optionName : s_optionNames) {
+            int len = strlen(optionName);
+            if (!option.compare(0, len, optionName)) {
```
You compare a string twice here and executeOption. It looks a bit nonsence.
if (option == "--resource-load-statistics-enabled") {
} else if (option == "--resource-load-statistics-disabled")

mokjasmine replied:
resolved in https://github.com/WebKit/WebKit/pull/5154/commits/ec72e16864b4383d8497aac77fa097414823a309

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +115,45 @@ MainWindow::MainWindow(const char* requestedURL)
+            requestedURL = option;
+    }
+
+    activeWebView()->loadURL("https://webkit.org");
```
Do you want to use `requestedURL` here?

mokjasmine replied:
resolved in https://github.com/WebKit/WebKit/pull/5154/commits/ec72e16864b4383d8497aac77fa097414823a309

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,9 +115,30 @@ MainWindow::MainWindow(const char* requestedURL)
 
+void MainWindow::executeOption(const String& option, String* requestedURL)
+{
+    if (!strcasecmp(option, s_resourceLoadStatisticsEnabled)) {
```
You should use `string::==` or `string::compare` here.

fujii replied:
Ah, you want to compare  strings case-insensitively. I don't think command switches should be case-insensitive. But, it's OK if you want.

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -41,7 +42,15 @@ constexpr int FontSize = 32;
-MainWindow::MainWindow(const char* requestedURL)
+const char* s_resourceLoadStatisticsEnabled = "--resource-load-statistics-enabled";
+const char* s_resourceLoadStatisticsDisabled = "--resource-load-statistics-disabled";
+
```
defining variables looks a bit nonsense here. You shouldn't define them.

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,9 +115,30 @@ MainWindow::MainWindow(const char* requestedURL)
     createNewWebView(nullptr);
+
+    // Parse command line options
+    String requestedURL = ""_s;
```
This is `WTF::String`. WTF isn't public API. Don't use WTF in MiniBrowser. use std::string instead.

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,9 +115,30 @@ MainWindow::MainWindow(const char* requestedURL)
     activeWebView()->loadURL(requestedURL ? requestedURL : "https://webkit.org");
 }
 
+void MainWindow::executeOption(const String& option, String* requestedURL)
```
I'd like to move the for loop into here and rename `executeOptions` .
 `executeOptions`  sounds a bit weird to me. How about `parseOptions`?

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,9 +115,30 @@ MainWindow::MainWindow(const char* requestedURL)
+        executeOption(option, &requestedURL);
+    }
+
     activeWebView()->loadURL(requestedURL ? requestedURL : "https://webkit.org");
```
requestedURL is initialised ""_s. Does this code work as expected? Can this conditional expression take the false expression?

donny-dont commented on Tools/MiniBrowser/playstation/MainWindow.h:
```diff
@@ -35,7 +35,7 @@ class WebViewWindow;
 class MainWindow final : public toolkitten::Widget {
 public:
-    MainWindow(const char*);
+    MainWindow(const vector<string>& options);
```
`std::vector`

donny-dont commented on Tools/MiniBrowser/playstation/MainWindow.h:
```diff
@@ -50,6 +50,7 @@ class MainWindow final : public toolkitten::Widget {
     void setActiveWebView(WebViewWindow*);
     void switchActiveWebView(int delta);
     void closeWebView(WebViewWindow*);
+    void executeOption(const String& option, String* requestedURL);
```
`const std::string&` for all I believe.

donny-dont commented on Tools/MiniBrowser/playstation/main.cpp:
```diff
@@ -93,7 +93,8 @@ int main(int argc, char *argv[])
     app.init(&applicationClient);
 
-    auto mainWindow = std::make_unique<MainWindow>(argc > 1 ? argv[1] : nullptr);
+    const vector<string> options(argv + 1, argv + argc);
```
`std::vector`

donny-dont reviewed (commented):
Fix up these and @fujii 's concerns and will do another review

donny-dont commented on Tools/MiniBrowser/playstation/MainWindow.h:
```diff
@@ -50,6 +51,7 @@ class MainWindow final : public toolkitten::Widget {
     void setActiveWebView(WebViewWindow*);
     void switchActiveWebView(int delta);
     void closeWebView(WebViewWindow*);
+    void parseOptions(const std::vector<std::string>& options, std::string* requestedURL);
```
Could we just have it return a `std::string` instead of doing the pointer to `std::string`?

fujii replied:
How about an idea of initializing `requestedURL` as "https://webkit.org", and pass the reference of the string to parseOptions?

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +109,27 @@ MainWindow::MainWindow(const char* requestedURL)
+
+    parseOptions(options, &requestedURL);
+
+    activeWebView()->loadURL(strlen(requestedURL.c_str()) ? requestedURL.c_str() : "https://webkit.org");
```
Don't use `strlen` here. Use size or length. https://en.cppreference.com/w/cpp/string/basic_string/size

fujii replied:
string::empty seems better. https://en.cppreference.com/w/cpp/string/basic_string/empty

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +109,27 @@ MainWindow::MainWindow(const char* requestedURL)
+{
+    auto context = WebContext::singleton();
+    WKWebsiteDataStoreRef websiteDataStore = context->websiteDataStore();
+    for (auto option : options) {
```
"auto" shold be "auto&" here. "auto" creates a string copy just for iterating.

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +109,27 @@ MainWindow::MainWindow(const char* requestedURL)
+    activeWebView()->loadURL(!requestedURL.empty() ? requestedURL.c_str() : "https://webkit.org");
+}
+
+void MainWindow::parseOptions(const std::vector<std::string>& options, std::string* requestedURL)
```
This `requestedURL` should never be nullptr because you dereference it unconditionally. In such case, you should use "std::string&".

fujii commented on Tools/MiniBrowser/playstation/MainWindow.cpp:
```diff
@@ -106,7 +109,27 @@ MainWindow::MainWindow(const char* requestedURL)
-    activeWebView()->loadURL(requestedURL ? requestedURL : "https://webkit.org");
+
+    // Parse command line options
+    std::string requestedURL = "";
```
How about an idea of initializing `requestedURL` as "https://webkit.org"?

fujii reviewed (approved):
LGTM.
Nit: There is a typo`dissable` in the commit message. And, the commit message looks old. It says `executeOption`.

mokjasmine committed:
Add command line option to enable/disable ResourceLoadStatistics when launching MiniBrowser
https://bugs.webkit.org/show_bug.cgi?id=246233

Reviewed by Fujii Hironori.

Use the options --resource-load-statistics-enabled and --resource-load-statistics-disabled to enable and disable ResourceLoadStatistics respectively. It is disabled by default.

* Tools/MiniBrowser/playstation/main.cpp:
(main):
* Tools/MiniBrowser/playstation/MainWindow:
(MainWindow::MainWindow):
(MainWindow::parseOptions):
* Tools/MiniBrowser/playstation/MainWindow.h:

Canonical link: https://commits.webkit.org/255703@main
